src/iot_network_routing/core/node.py

```
@@ -52,47 +124,43 @@ def add_neighbor(self, neighbor: 'IoTNode') -> bool:
         Returns:
             True if neighbor was added, False if out of range or already exists
         """
-        if neighbor == self:
+        if neighbor.eui64 == self.eui64:
             return False
             
         if not self.can_communicate_with(neighbor):
             return False
+        
+        # Check if edge already exists
+        if self._graph.has_edge(self.eui64, neighbor.eui64):
+            return False
             
-        # Add bidirectional connection
-        added = False
-        if neighbor not in self.neighbors:
-            self.neighbors.append(neighbor)
-            added = True
-            
-        if self not in neighbor.neighbors:
-            neighbor.neighbors.append(self)
-            added = True
-            
-        return added
+        # Add bidirectional edge with distance as weight
+        distance = self.distance_to(neighbor)
+        self._graph.add_edge(self.eui64, neighbor.eui64, weight=distance)
+        return True
     
     def remove_neighbor(self, neighbor: 'IoTNode') -> bool:
-        """Remove a neighbor from the list. Removes bidirectional connection."""
-        removed = False
-        if neighbor in self.neighbors:
-            self.neighbors.remove(neighbor)
-            removed = True
-            
-        if self in neighbor.neighbors:
-            neighbor.neighbors.remove(self)
-            removed = True
-            
-        return removed
+        """Remove a neighbor from the graph. Removes bidirectional connection."""
+        if self._graph.has_edge(self.eui64, neighbor.eui64):

```

defensive, should raise instead of return bool.